### ISSUE TITLE
Synchronize tags as well

### ISSUE THREAD

**Issue by xXSTrikeXx** [nextcloud/desktop#206]
Hello!

Could it be possible to Tag files in Windows explorer, like it is implemented with the shared option?
It will be a better workflow, because visiting all time the WebUI is annoying when you sometimes just want to flag/tag a file.

My first issue about that is in the wrong repo :D https://github.com/nextcloud/client_theming/issues/235
Cheers ;)

**Comment by camilasan** [comment]
Yes, this would be a nice feature.

**Comment by Phoara** [comment]
Any news on this feature? I would like to use Leap on my Mac which tags files using OpenMetaTags. When I add tags the files aren't synced though.

**Comment by vbspam** [comment]
I would also love to use it. At this moment I do not let Nextcloud Desktop client touch my tagged files and have to synchronize them through rsync.

**Comment by jcrXO** [comment]
I would love that feature too!

**Comment by elevatingcreativity** [comment]
I would like this as well, primarily for syncing with MacOS. Right now there is data loss, we use tags extensively in the native OS, but Nextcloud loses them all upon sync.

**Comment by mylos** [comment]
Just moved everything from dropbox to nextcloud as I realised the tags doesn't work and had to move everything back :(. Looking forward to see tags working to start using nextcloud! Thank you!

**Comment by meyerjom** [comment]
This would also be an absolutely important feature for me. It would be best, if you could maintain the tags, which are assigned on MacOS, directly in Nextcloud and the feature would be operating system independent.

**Comment by livier** [comment]
I want that feature too ...

**Comment by skarados** [comment]
Please add this feature

**Comment by JayBrown** [comment]
Yeah, extended attributes are super-important nowadays. Lots of information is in there. For example, the FOSS PDF reader Skim saves PDF notes in XAs, now also with the `#S` suffix for "syncable" metadata. Clouds need to to support metadata & XAs, at the very least those with the `#S` suffix (on macOS). Modern cloud services need that feature… simple as that.

Best approach would be a metadata config the user can populate with the names of the XAs he wants synchronized. So shite like `com.apple.quarantine` can be ignored, but Finder tags, 3rd-party XAs etc. can be synched.

**Comment by nivoc** [comment]
Would live this too. Sync between MacOS Tags and nextcloud Tags would be very nice!

**Comment by kkandert** [comment]
Yes please add sync support for metadata. 👍
Something like intelligent folders like on the mac would also be useful.

**Comment by tim-rohrer** [comment]
I had started exploring how to do with initially in PHP using `occ` command line, but then realized that in order to do this right, extended attributes like tags must be synchronized through the client apps. Unfortunately, C++/ObjC are not in my skill sets. 

I can +1 this enhancement request to death and ready & willing to test someone's work.

**Comment by livier** [comment]
It should work on W$, Mac and Linux, 
on clients, and web interface for the three 
I don't even know if the 3 Os system are compatible for tags admin. 
So, it is just a question : 
Is there really any hope to get that ?

**Comment by fknittel** [comment]
I've posted [a bounty on this feature](https://www.bountysource.com/issues/55875935-synchronize-tags-as-well). Maybe others would like to contribute too?

**Comment by NeutroGe** [comment]
Hey guys,

I've found a workaround until this is supported by nextcloud.

1°) install brew on all osx clients (https://brew.sh/)

2°) install the app "tag" (https://github.com/jdberry/tag) on all osx clients from terminal using the command "brew install tag"

3°) On the source osx client (the one where the files have tags), use the terminal to move inside the nextcloud folder then run the following command:

echo '#!/bin/sh' > result.sh ; find . -not -path '*/\.*' -exec printf 'tag -s '  \; -exec tag -N -0 {} \; -exec printf " " \; -exec tag -T -0 {} \; -exec echo "" \; > result.sh

This will generate a file named "result.sh".

On the target osx clients where the tags need to be set, simply execute the file result.sh and the tags will be set identically as the source osx client.

Both tasks can be added as cron jobs to automate the process on a regular basis so it's completely transparent to the end user.

**Comment by NeutroGe** [comment]
Hey guys,

Until this is implemented i found a workaround: exporting osx tags at the end of the files names.

I've created a small osx app which does that on the selected folder, you can download it here:

https://mirr.re/d/1bI4

This app requires the utility "tag" to work: https://github.com/jdberry/tag

Enjoy :)

**Comment by Boki4d** [comment]
> I've posted [a bounty on this feature](https://www.bountysource.com/issues/55875935-synchronize-tags-as-well). Maybe others would like to contribute too?

I'm in! This would be a really useful future, but it has to be cross-plattform (win, linux and macos).

**Comment by GlassedSilver** [comment]
Chipped in 15 bucks as well :)

**Comment by jampy** [comment]
Added to the bounty as well.

Could someone from the team like @DominiqueFuchs or @misch7  please give some feedback?

**Comment by GlassedSilver** [comment]
Any update on this?

**Comment by NeutroGe** [comment]
I have developped a small osx app that reads file tags and put them in the file name of the folder where it's ran on. This acts as a workaround for nextcloud not dealing (and probably never will) with osx tags.

If someone is interested i'll share it here later.

**Comment by fool15982** [comment]
> I have developped a small osx app that reads file tags and put them in the file name of the folder where it's ran on. This acts as a workaround for nextcloud not dealing (and probably never will) with osx tags.
> 
> If someone is interested i'll share it here later.

Is there a chance to get a copy?

**Comment by soerpoers** [comment]
add a bounty too – would be great if this feature comes around…

**Comment by NeutroGe** [comment]
> > I have developped a small osx app that reads file tags and put them in the file name of the folder where it's ran on. This acts as a workaround for nextcloud not dealing (and probably never will) with osx tags.
> > If someone is interested i'll share it here later.
> 
> Is there a chance to get a copy?

https://nextcloud.ah1z.com/s/L4yXjTEFs8wAiog

For it to work you need to install brew on osx, then from brew install the apps "tag" and "rename".

Then put a file with tags in a folder, put the app in it and run it, it will put the tags at the end of the filename.

The app is just this command line running when you launch it:

find . -not -path "*/*.*/*" -not -path "." -not -path "*/.*" -exec sh -c 'result=$(/usr/local/bin/tag -N "{}") ; if [ ! -z "$result" ]; then /usr/local/bin/rename -X -a" #$result" "{}"; fi' \; > /dev/null 2>&1

**Comment by trvrcr** [comment]
Is this in the pipeline?

**Comment by FlexW** [comment]
As far as I'm aware no one is working on it and it's not scheduled by now? @camilasan @er-vin @allexzander

**Comment by er-vin** [comment]
Indeed, it is not scheduled. Also it's a whole can of worms regarding integrating this into platforms... there would be at least four different system APIs to talk to IIRC.

**Comment by PechenTreski** [comment]
@er-vin maybe at least the webdav could be extended? I would like to download tags (that I can set in the web UI) via API. The background is: I have many scanned PDF documents in different languages and I need to OCR'ecognise them for a faster search via macos spotlight. For that I need to set tags with language codes like "eng", "deu", "rus". So my idea is to get that tags somehow and process the files according to tag values. 
Right now I don't see tag props in the webdav https://docs.nextcloud.com/server/latest/developer_manual/client_apis/WebDAV/search.html#supported-dav-properties

**Comment by NeutroGe** [comment]
> Is there an update on this issue? I use Nextcloud as my backup. After a hardware change all my Tags are gone!!!

This will never get fixed because it's too complicated to implement. You can use the workaround i posted earlier to put tags into filenames. As far as i know even onedrive or gsync dont have this feature. The only one that does is dropbox. But using it kinda defeats the whole purpose of using nextcloud so there are no perfect solution right now.

**Comment by GlassedSilver** [comment]
> > Is there an update on this issue? I use Nextcloud as my backup. After a hardware change all my Tags are gone!!!
> 
> This will never get fixed because it's too complicated to implement. You can use the workaround i posted earlier to put tags into filenames. As far as i know even onedrive or gsync dont have this feature. The only one that does is dropbox. But using it kinda defeats the whole purpose of using nextcloud so there are no perfect solution right now.

How can we be so sure it's too difficult to implement? The comparison with Google and Microsoft is a bit weak if you ask me.
The reason why a feature isn't present in some high-tier software often isn't a question of complexity or at least just that.

**Comment by NeutroGe** [comment]
> > > Is there an update on this issue? I use Nextcloud as my backup. After a hardware change all my Tags are gone!!!
> > 
> > 
> > This will never get fixed because it's too complicated to implement. You can use the workaround i posted earlier to put tags into filenames. As far as i know even onedrive or gsync dont have this feature. The only one that does is dropbox. But using it kinda defeats the whole purpose of using nextcloud so there are no perfect solution right now.
> 
> How can we be so sure it's too difficult to implement? The comparison with Google and Microsoft is a bit weak if you ask me. The reason why a feature isn't present in some high-tier software often isn't a question of complexity or at least just that.

Look at the previous answer by @er-vin in this thread: he says "it's a can of worm" because it is very complex, and he's a nextcloud dev. Also people who use tags are probably 0.5% of everyone using nextcloud so investing this much time into something a few people use is probably not worth it for them. Seeing for how long this thread has been opened and no fix has come we can assume it will never be fixed. Hey, i wish it would too but i'm just beeing realistic. Even when people offer money to fix it no one did, which means probably a few people already looked into it but gave up because too complicated.

**Comment by GlassedSilver** [comment]
> I'd somewhat like this feature, took a _very quick_ look into it, seemed like too big an investment. Seems unlikely this will happen, sadly

If your look into it was very quick by your own admission and given the importance of metadata around files these days I doubt you can really make a valid judgement on whether this is too big an investment. No hard feelings.

I've invested SO MUCH TIME in my macOS years in tags, coloring the tags, structuring smart folders (saved searches) around tags, etc.., all of that went down the drain when I went to Windows. The data is simply lost and my capabilities to sort files beyond the folder hierarchy is somewhat limited cross-platforms it's ridiculous.

There is a GREAT deal of UX improvement in this if implemented.

**Comment by S0ulf3re** [comment]
> > I'd somewhat like this feature, took a _very quick_ look into it, seemed like too big an investment. Seems unlikely this will happen, sadly
> 
> With all due respect, if your look into it was very quick by your own admission, given the importance of metadata around files these days, I doubt you can really make a valid judgement on whether this is too big an investment. No hard feelings.
> 
> I've invested SO MUCH TIME in my macOS years in tags, coloring the tags, structuring smart folders (saved searches) around tags, etc…, all of that went down the drain when I went to Windows. The data is simply lost and my capabilities to sort files beyond the folder hierarchy is somewhat limited cross-platforms it's ridiculous.
> 
> There is a GREAT deal of UX improvement in this if implemented.

Same here, I love the preservation that metadata brings. And as someone who is switching from iCloud to Nextcloud, I would love to have a seamless transition.

**Comment by kaystrobach** [comment]
End user perspective:

* Windows: https://www.lifewire.com/tag-files-in-windows-4587240
* Mac: https://www.techrepublic.com/article/how-to-use-tags-in-macos-finder/

Sadly I don't know how to have a good solution for linux yet.

**Comment by tim-rohrer** [comment]
@kaystrobach, I don't understand what these links have to do with NetCloud? Or are you just trying to show how simple it can be? 

Some Linux filesystems (if I recall correctly, but it was too long ago that I researched so I don't remember details) will support tags. But yeah, it was less than ideal as I recall. 

@NeutroGe, when you say, "people who use tags are probably 0.5% of everyone using nextcloud" you're missing the likely much larger percentage (like me) who evaluated NetCloud and wouldn't shift because of the lack of tag support. It is why I tried DropBox for a while, but when Apple released shared group folders for iCloud, I went there. I'd like to use NetCloud for various reasons, but without improved tag support, it isn't going to happen. 

I appreciate the sentiment that this support would be difficult, but I don't see it as impossible. Frankly, I see two high-level options:

1. Create and implement a fully supported tagging construct in NetCloud that works with Desktop and mobile, and can be managed for groups, and then provide a converter from macOS's tagging system.
2. Seek cross platform consensus about how to handle meta data between platforms' files systems.

Tbh, it has been a couple of years since I looked at this stuff. I had considered taking a stab at number 1 above, but since I had no real skills in mobile development or the languages being used, I didn't think it was worth my time.

**Comment by kaystrobach** [comment]
@tim-rohrer yes it was only for informational purpose of the client view.

**Comment by NeutroGe** [comment]
> @kaystrobach, I don't understand what these links have to do with NetCloud? Or are you just trying to show how simple it can be?
> 
> Some Linux filesystems (if I recall correctly, but it was too long ago that I researched so I don't remember details) will support tags. But yeah, it was less than ideal as I recall.
> 
> @NeutroGe, when you say, "people who use tags are probably 0.5% of everyone using nextcloud" you're missing the likely much larger percentage (like me) who evaluated NetCloud and wouldn't shift because of the lack of tag support. It is why I tried DropBox for a while, but when Apple released shared group folders for iCloud, I went there. I'd like to use NetCloud for various reasons, but without improved tag support, it isn't going to happen.
> 
> I appreciate the sentiment that this support would be difficult, but I don't see it as impossible. Frankly, I see two high-level options:
> 
> 1. Create and implement a fully supported tagging construct in NetCloud that works with Desktop and mobile, and can be managed for groups, and then provide a converter from macOS's tagging system.
> 2. Seek cross platform consensus about how to handle meta data between platforms' files systems.
> 
> Tbh, it has been a couple of years since I looked at this stuff. I had considered taking a stab at number 1 above, but since I had no real skills in mobile development or the languages being used, I didn't think it was worth my time.

It's Nextcloud, not Netcloud

**Comment by BJKle** [comment]
Can we expect an implementation? I’d like to move my files away from iCloud, but without this feature ☹️

**Comment by kaystrobach** [comment]
interesting for [@schullogin](https://github.com/Schullogin/) an Enterprise customer as well.

**Comment by Tekl** [comment]
I'm also looking for this feature (for macOS/iOS). Does the Bounty have to reach $250.000 to get the function implemented?

**Comment by JuliaKirschenheuter** [comment]
@kaystrobach please do not post a private support tickets in open GitHub issue. Thank you!

**Comment by GlassedSilver** [comment]
> I'm also looking for this feature (for macOS/iOS). Does the Bounty have to reach $250.000 to get the function implemented?

The bounty is SITTING at $250 and has been for a good while now.

Since this feature is likely to require at least two devs working on this who each specialize in (a) given platform(s) the value is probably too low, so if you feel like your contribution could propel this further go ahead, it's probably well invested.

I'm considering upping my stake as well, however since I've already been invested for a while I'd probably like to see some dev confirm they are even considering taking this on first, because I can't justify endlessly locking up capital without some idea of possible dev interest.

**Comment by kaystrobach** [comment]
Submitted an enterprise support request regarding the feature right now.

**Comment by BJKle** [comment]
Hi @Runtemund,
do you know the status of your pull request? It's been 3 months already.
Thank you for all your effort.

**Comment by Runtemund** [comment]
Hi @BJKle,
yes, I got a reply some weeks ago, but unfortunately I am heady loaded on work currently, so I need a little time to work on the requestet unit test for that.

**Comment by i2h3** [comment]
I just wanted to let you know that we have discussed this topic yesterday again. There are some catches, pitfalls and caveats in the multi-platform and server-centric environment we are working in. One of the rather simple examples: macOS does not support custom tag colors. Hence it is impossible to represent tags from Nextcloud server correctly. We leave it for now because it is not that much of a pressing issue (compared to others).

**Comment by GlassedSilver** [comment]
@i2h3 are you sure about this? IIRC I set up tags with colors in Finder with custom colors or do you mean you're limited to a few pre-defined colors? In any case if there was a way to sync the tags and set the tag color at the OS-level on macOS that'd already be a big win in my books, because tags have become such a Godsend for working with files over recent years. 💕

**Comment by i2h3** [comment]
@GlassedSilver Yes. First: [the tag data on a file provider item](https://developer.apple.com/documentation/fileprovider/nsfileprovideritemprotocol/tagdata) is an opaque and abstract data blob not meant for parsing but storage as it is. Hence it is unreasonable to map for us already because of that. We can actually "hack" into it but without official support it could break any time Apple changes something internally in its framework which sets and gets that data blob. Besides that not being a reasonable offering to users and customers, the additional problem already is as suspected: colors are defined by an index in whatever Finder has as a list for colors.

 Xcode debugger output:

```plaintext
(lldb) po (try? NSKeyedUnarchiver.unarchivedObject(ofClasses: [NSArray.self, NSMutableArray.self, NSDictionary.self, NSString.self, NSNumber.self, NSData.self], from: item.tagData!! )) as Any
▿ Optional<Any>
  ▿ some : 2 elements
    ▿ 0 : 2 elements
      - key : v
      - value : 1
    ▿ 1 : 2 elements
      - key : t
      ▿ value : 1 element
        ▿ 0 : 2 elements
          - 0 : Super
          - 1 : 5
```

Swift literal representation:

```swift
[
    "v": 1,
    "t": [
        ("Super", 5)
    ]
]
```

**So, as of now, supporting Nextcloud tags in the form of mapping them to native macOS tags is not possible without substantial caveats and reliance on private APIs.**

Technically, with the addition of another BLOB property on file items, we could synchronize that tag data from Apple devices to devices consistently (but not to other supported client platforms) but that counteracts the server-centric and client platform-agnostic intent of Nextcloud.

**Comment by ismasans** [comment]
I’ve read this very old thread and, as far as I can make out (English isn’t a language I’m fluent in and I’m not a developer), the difficulty lies in making macOS tags compatible with, or combining them with, Windows tags and Nextcloud’s own tags.
Perhaps that approach is a mistake and it would be simpler to have these attributes synchronised between devices within the same family. In other words, Mac with Mac (or Apple), Windows with Windows, Linux with Linux... But not to try to make this feature cross-platform.
I’m not sure how complex it would be to implement, but it should be considered, especially if it’s true that, at the very least, we Mac users are asking for it, and it’s something that would make Nextcloud an ‘almost’ perfect service. ;-)

**Comment by GlassedSilver** [comment]
> I’ve read this very old thread and, as far as I can make out (English isn’t a language I’m fluent in and I’m not a developer), the difficulty lies in making macOS tags compatible with, or combining them with, Windows tags and Nextcloud’s own tags. Perhaps that approach is a mistake and it would be simpler to have these attributes synchronised between devices within the same family. In other words, Mac with Mac (or Apple), Windows with Windows, Linux with Linux... But not to try to make this feature cross-platform. I’m not sure how complex it would be to implement, but it should be considered, especially if it’s true that, at the very least, we Mac users are asking for it, and it’s something that would make Nextcloud an ‘almost’ perfect service. ;-)

IMHO both are rather reasonable needs.

On one hand synchronizing between whatever can be reliably made universal and also the blob-way of Apple, since I'd rather not LOSE the metadata of files created on Apple between compatible machines rather than experience SILENT metadata loss.

That being said, I think this feels like something Apple should be pressured to make open and for app developers to reliably tap into.

Thank you, @i2h3, for the great explanation by the way, that was an incredibly insightful deep-dive. :)
